### PR TITLE
FEXCore: Reclaimable thread pool allocator

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Creates a pool allocator for OpcodeDispatcher and IRCompaction that
shares memory allocations between threads in a pool and supports
reclaiming stale allocations from participating threads.

A thread will use a heuristic to keep its claimed memory allocation
around if it is allocating a lot of code. If it slows down then it will
start putting the memory allocation back in to the thread pool.

Additionally if the allocation has been "disowned" and gone to sleep
while still retaining the allocation, then another thread can inspect
 these stale allocations and reclaim it from the idling thread. Saving
further memory.

This needs some more work and cleanup but this is an interesting concept
that saves a decent amount of memory even in a basic test.

Causes teeworlds' title screen to go from 754MB to 599MB in my simple
test. 79.4% the memory usage is a good start.

First step for #1625

### CODE REVIEW

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+      struct MemoryBuffer {
+        void* Ptr;
+        size_t Size;
+        std::chrono::time_point<ClockType> LastUsed;
+        std::atomic<bool> ClientClaimed;
+        std::atomic<bool> *CurrentClientOwnedFlag{};
+      };
+
+      using ContainerType = std::list<MemoryBuffer*>;
```
Comment (neobrain) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:30]: Would `std::vector` work here? Looks like we're mostly just iterating over these (where `vector` provides better cache locality) and insertion/erasure is exceptional.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      void DisownBuffer(ContainerType::iterator Buffer) {
+        // Client still owns the buffer but isn't using it
+        // Allows us to claim it back if necessary
+        (*Buffer)->LastUsed = ClockType::now();
+        (*Buffer)->ClientClaimed.store(false);
+      }
+
+      ~ThreadPoolAllocator() {
+        // Cleanup any buffers that are remaining
+        for (auto it = Buffers.begin(); it != Buffers.end(); ++it) {
```
Comment (neobrain) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:54]: range-based for loops can be used here and in a couple of other places.

File: External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h [hunk 3]
```diff
@@ -20,19 +21,12 @@ namespace FEXCore::IR {
  *
  * Can potentially support reallocation if we are smart and make sure to invalidate anything holding a true pointer
  */
-class DualIntrusiveAllocator final {
+class DualIntrusiveAllocator {
```
Comment (neobrain) [inline External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:24]: Should this have a virtual destructor? If not, we should use non-public inheritance instead so people can't accidentally bind child objects to `DualIntrusiveAllocator*` pointers.

File: External/FEXCore/include/FEXCore/IR/IREmitter.h [hunk 4]
```diff
@@ -19,11 +19,37 @@ friend class FEXCore::IR::Pass;
...
+    IREmitter(FEXCore::Utils::ThreadPoolAllocator &ThreadAllocator)
+      : DualListData {ThreadAllocator, 8 * 1024 * 1024} {
       ResetWorkingList();
     }
 
+    void ClaimBuffer() {
+      DualListData.ClaimBuffer();
+    }
+    void UnclaimBuffer() {
+      auto Now = std::chrono::high_resolution_clock::now();
```
Comment (neobrain) [inline External/FEXCore/include/FEXCore/IR/IREmitter.h:31]: Would a `steady_clock` be sufficient here as well? A high resolution clock doesn't appear necessary given that the reclamation interval is 5 seconds, and `steady_clock` (being a monotonic clock) will have more predictable behavior on system clock changes. (You wouldn't want reclamation to stop working for an hour because the system entered DST!)

File: External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h [hunk 5]
```diff
@@ -20,19 +21,12 @@ namespace FEXCore::IR {
...
   public:
     DualIntrusiveAllocator() = delete;
```
Comment (neobrain) [inline External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:26]: Aren't these two constructors implicitly deleted due to providing the `size_t` constructor?

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 1 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:30]: I very explicitly chose list here so iterators aren't invalidated on insert/delete.
Could add a comment stating why I chose list.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 2 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Range based used for the ones that can. The other ones still need the iterators to be passed around.

File: External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h (hunk 3 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:24]: Yes. Resolved.

File: External/FEXCore/include/FEXCore/IR/IREmitter.h (hunk 4 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/IR/IREmitter.h]: You right. Resolved.

File: External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h (hunk 5 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h]: You right. Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 1 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:31]: Added a comment that I'm requiring iterators to stay valid while manipulating the container. Resolved.

**Comment by Sonicadvance1** [comment]
Fixed the remaining bugs in this and added a bunch of comments about how to use the API.
Now the most major contributor to memory consumption is the register allocator. Specifically the graph nodes since they aren't ever freed.

Once this is merged those can be looked at to fix its memory consumption, since it isn't quite as straightforward as just moving the allocator over to using these.

**Comment by Sonicadvance1** [comment]
Image_2022-04-24_11-30-29
Heaptrack shows a significant change but it isn't this significant because some allocations have switched from malloc to mmap. 1.6G down to 229.3MB within FEXLoader.
I don't think there is a tool that can see resident memory usage reduction other than what tools like htop give, sadly.

**Comment by skmp** [comment]
I'll get on reviewing this first thing tomorrow, sorry for the delay

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+  };
+
+  /**
+   * @brief Thread pool allocator that allocates and frees objects using malloc
+   */
+  class ThreadPoolAllocatorMalloc final : public IntrusiveThreadPoolAllocator {
+    public:
+      ThreadPoolAllocatorMalloc() = default;
+
+      virtual ~ThreadPoolAllocatorMalloc() {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:316]: Since both dtors `ThreadPoolAllocatorMalloc`, `ThreadPoolAllocatorMMap` are the same, wouldn't it be nicer to have the dtor logic in `IntrusiveThreadPoolAllocator` and rely on the alloc/free overloads? Unless I'm missing something

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+                  LRUTime = UsedTime;
+                }
+              }
+            }
+
+            // If we found a buffer then free it
+            if (LastUsed != BuffersEnd) {
+              Free((*LastUsed)->Ptr, (*LastUsed)->Size);
+              delete *LastUsed;
+              OwnedBuffers.erase(LastUsed);
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:254]: `it = OwnedBuffers.erase(LastUsed);` ?

Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:254]: This is no longer isn't the for loop, there is no iterator to update.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+              UnsizedFit = it;
+            }
+          }
+
+          // If we didn't have an exact fit then use an unsized fit
+          if (BestFit == BuffersEnd) {
+            BestFit = UnsizedFit;
+          }
+
+          {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:232]:
```suggestion
          // Free owned buffers that have expired
          {
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+                  it = NonOwnedBuffers.erase(it);
+                  continue;
+                }
+              }
+            }
+
+            ++it;
+          }
+        }
+
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:213]:
```suggestion
          // Find an owned buffer that is >= Size and Free owned buffers that have expired
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+    protected:
+      IntrusiveThreadPoolAllocator() = default;
+
+      ContainerType::iterator ClaimBufferImpl(size_t Size) {
+        auto BuffersEnd = OwnedBuffers.end();
+        ContainerType::iterator BestFit = BuffersEnd;
+        ContainerType::iterator UnsizedFit = BuffersEnd;
+
+        auto Now = ClockType::now();
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:188]:
```suggestion
        auto Now = ClockType::now();
        // Move any expired NonOwnedBuffers to OwnedBuffers
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+      void UnclaimBufferImpl(ContainerType::iterator Buffer) {
+        (*Buffer)->CurrentClientOwnedFlag = nullptr;
+        OwnedBuffers.emplace_back(*Buffer);
+        NonOwnedBuffers.erase(Buffer);
+      }
+
+      /**
+       * @brief List of buffers that this pool allocator itself owns
+       */
+      ContainerType OwnedBuffers;
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:279]: How about renaming to `UnclaimedBuffers`,  and `ClaimedBuffers` so that `ClaimBuffer`/`ClaimBufferImpl` & `UnclaimBuffer`/`UnclaimBufferImpl` are more intuitive to understand?

`DisownBuffer` and `ReownOrClaimBuffer` don't modify those lists, and from my understanding they track the `claimed`/`unclaimed` statuses, not `owned`/`disowned`

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+   *    - Updating two std::list containers to give the ownership back to the `Allocator`
+   *
+   *  - Claiming is very costly
+   *    - Requires owning a mutex, shared between all threads using the `Allocator`
+   *    - Scans two std::list containers to find the best fit buffer
+   *    - Or allocates another buffer when that fails
+   *    - Frees stale buffers opportunistically
+   */
+  template<typename Type, size_t PeriodMS, size_t PeriodFrequency>
+  class FixedSizeDelayedObjectReclaimer final {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:414]: How about we name this from `FixedSizeDelayedObjectReclaimer` to `FixedSizePooledAllocation` for clarity?

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+   *       - This is expensive and requires a mutex shared between threads
+   *     - `FixedSizeDelayedObjectReclaimer` helper class provided to help with this.
+   *
+   * Once the client has disowned a buffer then the allocator is free to reclaim the buffer when another thread is trying to `Claim` a new buffer.
+   * The buffer getting claimed from a disowned client must have had its last use greater than the defined `DURATION` before it has a chance to get
+   * reclaimed by the Allocator.
+   *
+   * During buffer reclaiming is also when unclaimed buffers get freed. This means active threads are able to clean up idle thread's unused memory.
+   */
+  class IntrusiveThreadPoolAllocator {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:35]: How about we rename this to  `IntrusivePooledAllocator` to make it clear it doesn't allocate from a `ThreadPool`?

**Comment by skmp** [comment]
Some performance points for discussion:

Switching `IntrusiveThreadPoolAllocator::ContainerType` to a multimap structure, then ordering owned buffers by size can make finding the right buffer O(ln) from O(n), possibly a major win depending on how many buffers we have in flight.

Not sure the same idea would be applicable for unowned buffers (sorted by last used time), as that'd require a re-insertion on disown which will increase the cost there.

What is the cost of `std::chrono` and such? Do they result in syscalls? Can we guarantee somehow `rdtsc`/`cntvct_el0` for it?

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 6 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:316]: Can't
`Call to pure virtual member function 'Free' has undefined behavior; overrides of 'Free' in subclasses are not available in the destructor of 'IntrusivePooledAllocator'  ThreadPoolAllocator.h:323:7: note: 'Free' declared here [-Wcall-to-pure-virtual-from-ctor-dtor]`

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 8 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:235]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 9 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:214]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 11 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 12 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 13 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

**Comment by Sonicadvance1** [comment]
`std::chrono` doesn't result in a syscall. It'll use the cycle counter registers, potentially through a vdso. quite lightweight.

I haven't noticed any performance concerns from the current `std::list` container but I've also not found something that pedantically disowns thousands of threads worth of buffers and then spins all of them up simultaneously so they are all hunting for buffers to reown.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 6 above)
Comment (skmp) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:319]: Doh, right, derived class is destructed at that point. How about moving them to a function so that we don't have two copies of the code around?

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 7 above)
Comment (skmp) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Misread the indentation. So at most one buffer gets free'd for every call to `ClaimBufferImpl`. Is this to avoid stutter?

Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Yes, only one unclaimed buffer gets freed and removed from the `UnclaimedBuffers` container each time a new claim comes in. This is setup this way for longer running applications having mostly idle threads and then a handful of active threads will slowly chip away at the problem.
Also if you have an intermediate mix of active and inactive, you aren't suddenly getting a spike of deleted unclaimed buffers and new threads having absolutely no buffers available.

Basically the five second buffer of only freeing one buffer at a time is a nice sponge in both directions.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 6 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:334]: Moved duplicated freeing logic to its own function. Resolved now.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+              UnsizedFit = it;
+            }
+          }
+
+          // If we didn't have an exact fit then use an unsized fit
+          if (BestFit == BuffersEnd) {
+            BestFit = UnsizedFit;
+          }
+
+          // Free owned buffers that have expired
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:234]:
```suggestion
          // Free up to one unclaimed buffer that has expired
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+                  continue;
+                }
+              }
+            }
+
+            ++it;
+          }
+        }
+
+        // Find an owned buffer that is >= Size and Free owned buffers that have expired
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:215]:
```suggestion
        // Find an unclaimed buffer that is >= Size and Free up to one unclaimed buffer that has expired
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+
+      void Free(void* Ptr, size_t Size) override {
+        FEXCore::Allocator::free(Ptr);
+      }
+  };
+
+  /**
+   * @brief Thread pool allocator that allocates and frees objects that uses mmap
+   */
+  class ThreadPoolAllocatorMMap final : public IntrusivePooledAllocator {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:351]:
```suggestion
  class PooledAllocatorMMap final : public IntrusivePooledAllocator {
```

or 

```suggestion
  class PooledMMapAllocator final : public IntrusivePooledAllocator {
```

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h [hunk 17]
```diff
@@ -0,0 +1,502 @@
...
+       * @param Ptr buffer pointer
+       * @param Size buffer size
+       */
+      virtual void Free(void* Ptr, size_t Size) = 0;
+  };
+
+  /**
+   * @brief Thread pool allocator that allocates and frees objects using malloc
+   */
+  class ThreadPoolAllocatorMalloc final : public IntrusivePooledAllocator {
```
Comment (skmp) [inline External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:330]:
```suggestion
  class PolledAllocatorMalloc final : public IntrusivePooledAllocator {
```

Or

```suggestion
  class PolledMallocAllocator final : public IntrusivePooledAllocator {
```
?

**Review by skmp — APPROVED** [review]
Noted a few naming nitpicks, Looks good overall

**Comment by skmp** [comment]
> `std::chrono` doesn't result in a syscall. It'll use the cycle counter registers, potentially through a vdso. quite lightweight.
> 
> I haven't noticed any performance concerns from the current `std::list` container but I've also not found something that pedantically disowns thousands of threads worth of buffers and then spins all of them up simultaneously so they are all hunting for buffers to reown.

More of a note for the future then

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 16 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 15 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

File: External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h (hunk 14 above)
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h]: Resolved.

File: External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h (hunk 3 above)
Comment (neobrain) [reply External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:24]: What happened to this? Looks like the destructor is gone again now.
